geode-core/src/integrationTest/java/org/apache/geode/cache/query/JoinQueriesIntegrationTest.java

```diff
@@ -120,4 +175,167 @@ private void populateRegionWithData(Region region1, Region region2) {
+      validationIssueXRefRegion.put("validationIssueXRefID_" + i, new OrderValidationIssueXRef(
+          "validationIssueXRefID_" + i, "issueId_" + i, "orderId_" + i, i));
+    }
+
+    for (int i = 0; i < extraEntitiesPerRegion; i++) {
+      orderRegion.put("orderId#" + i, new Order("orderId#" + i, i));
+      validationIssueRegion.put("referenceIssueId#" + i,
+          new ValidationIssue("referenceIssueId#" + i, Calendar.getInstance().getTime()));
+      validationIssueXRefRegion.put("validationIssueXRefID#" + i, new OrderValidationIssueXRef(
+          "validationIssueXRefID#" + i, "validationIssueID2#" + i, "orderId#2" + i, i));
+    }
+  }
+
+  @Test
+  @Parameters({"true", "false"})
+  @TestCaseName("{method} - Using PDX: {params}")
+  public void joiningThreeRegionsWhenIntermediateResultSizeIsHigherThanLimitClauseShouldNotTrimResults(
+      boolean usePdx) throws Exception {
+    int matches = 10;
+    int extraEntitiesPerRegion = 25;
+    InternalCache cache = serverRule.getCache();
+    QueryService queryService = cache.getQueryService();
+    String queryString = "SELECT issue.issueId, issue.createdTime, o.orderId, o.version "
+        + "FROM /ValidationIssue issue, /OrderValidationIssueXRef xRef, /Order o "
+        + "WHERE "
+        + "issue.issueId = xRef.referenceIssueId "
+        + "AND "
+        + "xRef.referenceOrderId = o.orderId "
+        + "AND "
+        + "xRef.referenceOrderVersion = o.version ";
+
+    Region<String, Object> orderRegion = cache.<String, Object>createRegionFactory()
+        .setDataPolicy(DataPolicy.REPLICATE).create("Order");
+    Region<String, Object> validationIssueRegion = cache.<String, Object>createRegionFactory()
+        .setDataPolicy(DataPolicy.REPLICATE).create("ValidationIssue");
+    Region<String, Object> validationIssueXRefRegion = cache.<String, Object>createRegionFactory()
+        .setDataPolicy(DataPolicy.REPLICATE).create("OrderValidationIssueXRef");
+
+    if (!usePdx) {
+      populateTripleJointRegions(matches, extraEntitiesPerRegion, orderRegion,
+          validationIssueRegion, validationIssueXRefRegion);
+    } else {
+      populateTripleJointRegions(matches, extraEntitiesPerRegion, cache, orderRegion,
+          validationIssueRegion, validationIssueXRefRegion);
+    }
+
+    SelectResults baseResults = (SelectResults) queryService.newQuery(queryString).execute();
+    SelectResults resultsWithLimitOne =
+        (SelectResults) queryService.newQuery(queryString + "LIMIT 2").execute();
+    SelectResults resultsWithLimitTwo =
+        (SelectResults) queryService.newQuery(queryString + "LIMIT 5").execute();
+    assertThat(baseResults.size()).isEqualTo(matches);
+    assertThat(resultsWithLimitOne.size()).isEqualTo(2);
+    assertThat(resultsWithLimitTwo.size()).isEqualTo(5);
+
+    // Add Indexes and test again
+    cache.getQueryService().createIndex("order_orderID", "orderId", "/Order", null);
```

Comment:
Maybe separate index vs non index into a separate test?  That way when one fails it will be easier to spot?